{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "77fcb18c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import random\n",
    "import csv\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a058a841",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chunk_sequence(seq, chunk_size=512, stride=256):\n",
    "    seq = seq.upper()\n",
    "    for i in range(0, len(seq) - chunk_size + 1, stride):\n",
    "        yield seq[i:i + chunk_size]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "03b87bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_sequence(seq):\n",
    "    seq = seq.upper()\n",
    "    seq = re.sub(r\"[^ACGTN]\", \"\", seq)\n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b0b18e37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasta_to_csv(fasta_path, csv_writer, label, name, source=\"GenBank\", min_len=200 ):\n",
    "    count = 0\n",
    "    for record in SeqIO.parse(fasta_path, \"fasta\"):\n",
    "        seq = clean_sequence(str(record.seq))\n",
    "        if len(seq) < min_len:\n",
    "            continue\n",
    "\n",
    "        seq_id = record.id\n",
    "\n",
    "        csv_writer.writerow([\n",
    "            seq_id,\n",
    "            name,\n",
    "            seq,\n",
    "            label,\n",
    "            source,\n",
    "            len(seq)\n",
    "        ])\n",
    "        count += 1\n",
    "\n",
    "    print(f\"[✓] {count} séquences enregistrees \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c40cc9e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_dataset(fasta_path, output_csv, isGMO=False):\n",
    "    file_exists = os.path.exists(output_csv)\n",
    "    os.makedirs(os.path.dirname(output_csv), exist_ok=True)\n",
    "    \n",
    "    filename = os.path.basename(fasta_path)\n",
    "    name = os.path.splitext(filename)[0]\n",
    "    seqName = \"_\".join(name.split(\"_\")[:2])\n",
    "\n",
    "    with open(output_csv, \"a\", newline=\"\") as f:\n",
    "        writer = csv.writer(f)\n",
    "        if not file_exists:\n",
    "            writer.writerow([\"id\", \"name\", \"sequence\", \"label\", \"source\", \"length\"])\n",
    "\n",
    "        fasta_to_csv(\n",
    "            fasta_path=fasta_path,\n",
    "            csv_writer=writer,\n",
    "            label= 1 if isGMO else 0,\n",
    "            name=seqName\n",
    "        )\n",
    "\n",
    "    print(f\"[✓] Dataset Enregistree vers {output_csv}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "17b623bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓] 1027 séquences enregistrees \n",
      "[✓] Dataset Enregistree vers processed/datas.csv\n"
     ]
    }
   ],
   "source": [
    "hirsutum_fasta=\"raw/Gossypium_hirsutum_v2.1_genomic.fna\"\n",
    "output_csv=\"processed/datas.csv\"\n",
    "build_dataset(fasta_path=hirsutum_fasta, output_csv=output_csv, isGMO=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ff49c925",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓] 676 séquences enregistrees \n",
      "[✓] Dataset Enregistree vers processed/datas.csv\n"
     ]
    }
   ],
   "source": [
    "unguiculata_fasta=\"raw/Vigna_unguiculata_v2_genomic.fna\"\n",
    "output_csv=\"processed/datas.csv\"\n",
    "build_dataset(fasta_path=unguiculata_fasta, output_csv=output_csv, isGMO=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2adf9d62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓] 20 séquences enregistrees \n",
      "[✓] Dataset Enregistree vers processed/datas.csv\n"
     ]
    }
   ],
   "source": [
    "cry1ac_fasta=\"raw/bt_cry1Ac_sequences.fasta\"\n",
    "output_csv=\"processed/datas.csv\"\n",
    "build_dataset(fasta_path=cry1ac_fasta, output_csv=output_csv, isGMO=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e2614447",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓] 11 séquences enregistrees \n",
      "[✓] Dataset Enregistree vers processed/datas.csv\n"
     ]
    }
   ],
   "source": [
    "cp4epsps_fasta=\"raw/bt_cp4epsps_sequences.fasta\"\n",
    "output_csv=\"processed/datas.csv\"\n",
    "build_dataset(fasta_path=cp4epsps_fasta, output_csv=output_csv, isGMO=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cb27efa7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓] 3 séquences enregistrees \n",
      "[✓] Dataset Enregistree vers processed/datas.csv\n"
     ]
    }
   ],
   "source": [
    "cotton_gmo_fasta=\"raw/cotton_gmo.fna\"\n",
    "output_csv=\"processed/datas.csv\"\n",
    "build_dataset(fasta_path=cotton_gmo_fasta, output_csv=output_csv, isGMO=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d39814c",
   "metadata": {},
   "source": [
    "Synthetisation de GMO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "84b62b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean(seq):\n",
    "    seq = seq.upper()\n",
    "    return re.sub(r\"[^ACGT]\", \"\", seq)\n",
    "\n",
    "def random_subseq(seq, min_len, max_len):\n",
    "    L = len(seq)\n",
    "    size = random.randint(min_len, max_len)\n",
    "    if L <= size:\n",
    "        return seq\n",
    "    start = random.randint(0, L - size)\n",
    "    return seq[start:start+size]\n",
    "\n",
    "def mutate(seq, rate=0.001):\n",
    "    bases = [\"A\",\"C\",\"G\",\"T\"]\n",
    "    seq = list(seq)\n",
    "    for i in range(len(seq)):\n",
    "        if random.random() < rate:\n",
    "            seq[i] = random.choice(bases)\n",
    "    return \"\".join(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7eea412c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_synthetic_gmo(plant_genomes, promoter_seqs, gene_seqs, terminator_seqs,\n",
    "        n_samples=1000, flank_min=300, flank_max=2000, mutation_rate=0.0005):\n",
    "\n",
    "    samples = []\n",
    "\n",
    "    for i in range(n_samples):\n",
    "        plant = random.choice(plant_genomes)\n",
    "        promoter = random.choice(promoter_seqs)\n",
    "        gene = random.choice(gene_seqs)\n",
    "        terminator = random.choice(terminator_seqs)\n",
    "\n",
    "        left_flank  = random_subseq(plant, flank_min, flank_max)\n",
    "        right_flank = random_subseq(plant, flank_min, flank_max)\n",
    "\n",
    "        synthetic = (left_flank + promoter + gene + terminator + right_flank)\n",
    "\n",
    "        synthetic = mutate(synthetic, mutation_rate)\n",
    "        \n",
    "        samples.append({\n",
    "            # \"id\": f\"GMO_hisu_SYN_{i:06d}\",\n",
    "            \"id\": f\"GMO_ungui_SYN_{i:06d}\",\n",
    "            \"name\" : \"synthetic_GMO\",\n",
    "            \"sequence\": synthetic,\n",
    "            \"label\": 1,\n",
    "            \"source\": \"me\",\n",
    "            \"length\": len(synthetic)\n",
    "        })\n",
    "\n",
    "    return samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "603bcba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def load_fasta(path):\n",
    "    seqs = []\n",
    "    for rec in SeqIO.parse(path, \"fasta\"):\n",
    "        seqs.append(clean(str(rec.seq)))\n",
    "    return seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8cd475db",
   "metadata": {},
   "outputs": [],
   "source": [
    "def append_to_csv(samples, csv_path):\n",
    "    file_exists = os.path.exists(csv_path)\n",
    "    os.makedirs(os.path.dirname(csv_path), exist_ok=True)\n",
    "\n",
    "    with open(csv_path, \"a\", newline=\"\") as f:\n",
    "        writer = csv.writer(f)\n",
    "        if not file_exists:\n",
    "            writer.writerow([\"id\", \"name\", \"sequence\", \"label\", \"source\", \"length\"])\n",
    "\n",
    "        for s in samples:\n",
    "            writer.writerow([s[\"id\"], s[\"name\"], s[\"sequence\"], s[\"label\"], s[\"source\"], s[\"length\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d29923c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plant_genomes = load_fasta(\"raw/Gossypium_hirsutum_v2.1_genomic.fna\")\n",
    "plant_genomes = load_fasta(\"raw/Vigna_unguiculata_v2_genomic.fna\")\n",
    "promoters     = load_fasta(\"raw/promoters.fasta\")\n",
    "genes         = load_fasta(\"raw/gmo_genes.fasta\")\n",
    "terminators   = load_fasta(\"raw/terminators.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4c55050a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓] Generated 5000 synthetic GMO sequences\n"
     ]
    }
   ],
   "source": [
    "synthetic_samples = build_synthetic_gmo(plant_genomes=plant_genomes, promoter_seqs=promoters, gene_seqs=genes,\n",
    "    terminator_seqs=terminators, n_samples=5000, flank_min=500, flank_max=3000, mutation_rate=0.0003)\n",
    "\n",
    "append_to_csv(synthetic_samples, csv_path=\"processed/datas.csv\")\n",
    "\n",
    "print(f\"[✓] Generated {len(synthetic_samples)} synthetic GMO sequences\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40cac2eb",
   "metadata": {},
   "source": [
    "Subdivision du dataset en train, test et validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54cd330b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "156d6c9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_dataset(input_csv, output_dir, train_ratio=0.7, val_ratio=0.2, test_ratio=0.1, random_state=42):\n",
    "    assert abs(train_ratio + val_ratio + test_ratio - 1.0) < 1e-6\n",
    "\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    df = pd.read_csv(input_csv)\n",
    "\n",
    "    # Premier decoupage: train et temp (val+test)\n",
    "    train_df, temp_df = train_test_split(df, test_size=(1 - train_ratio), stratify=df[\"label\"], random_state=random_state)\n",
    "\n",
    "    # Second decoupage: val vs test\n",
    "    val_size = val_ratio / (val_ratio + test_ratio)\n",
    "\n",
    "    val_df, test_df = train_test_split(\n",
    "        temp_df,\n",
    "        test_size=(1 - val_size),\n",
    "        stratify=temp_df[\"label\"],\n",
    "        random_state=random_state\n",
    "    )\n",
    "\n",
    "    train_df.to_csv(os.path.join(output_dir, \"train.csv\"), index=False)\n",
    "    val_df.to_csv(os.path.join(output_dir, \"val.csv\"), index=False)\n",
    "    test_df.to_csv(os.path.join(output_dir, \"test.csv\"), index=False)\n",
    "\n",
    "    print(\"[✓]Dataset split completed:\")\n",
    "    print(f\"  Train: {len(train_df)}\")\n",
    "    print(f\"  Val:   {len(val_df)}\")\n",
    "    print(f\"  Test:  {len(test_df)}\")\n",
    "\n",
    "    print(\"\\nLabel distribution:\")\n",
    "    print(\"Train:\\n\", train_df[\"label\"].value_counts(normalize=True))\n",
    "    print(\"Val:\\n\", val_df[\"label\"].value_counts(normalize=True))\n",
    "    print(\"Test:\\n\", test_df[\"label\"].value_counts(normalize=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "84d2c9bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓]Dataset split completed:\n",
      "  Train: 8215\n",
      "  Val:   2347\n",
      "  Test:  1175\n",
      "\n",
      "Label distribution:\n",
      "Train:\n",
      " label\n",
      "1    0.8549\n",
      "0    0.1451\n",
      "Name: proportion, dtype: float64\n",
      "Val:\n",
      " label\n",
      "1    0.854708\n",
      "0    0.145292\n",
      "Name: proportion, dtype: float64\n",
      "Test:\n",
      " label\n",
      "1    0.855319\n",
      "0    0.144681\n",
      "Name: proportion, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "split_dataset(input_csv=\"processed/datas.csv\", output_dir=\"processed/splits\", \n",
    "        train_ratio=0.7, val_ratio=0.2, test_ratio=0.1, random_state=42)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "409fa28a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[✓] Aucune fuite détectée\n"
     ]
    }
   ],
   "source": [
    "train_ids = set(pd.read_csv(\"processed/splits/train.csv\")[\"id\"])\n",
    "val_ids   = set(pd.read_csv(\"processed/splits/val.csv\")[\"id\"])\n",
    "test_ids  = set(pd.read_csv(\"processed/splits/test.csv\")[\"id\"])\n",
    "\n",
    "assert train_ids.isdisjoint(val_ids)\n",
    "assert train_ids.isdisjoint(test_ids)\n",
    "assert val_ids.isdisjoint(test_ids)\n",
    "print(\"[✓] Aucune fuite détectée\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "839a9639",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
